pre-commit integration v2

AleksaC opened the pull request:
Here's an improved version of pre-commit integration. I think it's in accordance with what @janosh and @asottile proposed in #765 at least in terms of interface. I made a [test repo](https://github.com/AleksaC/jsonschema-test-pre-commit) with pre-commit github action where you can check how the hook works on a simple test example. Any feedback is welcome.

janosh commented:
Great work!

Is it necessary to release this hook with a new id (`jsonschema-v2`)? Perhaps its release can coincide with a major version bump of `jsonschema` so users know to expect breaking changes? Of course if none is planned in the foreseeable future, this is impractical.

AleksaC commented:
Here's a fresh parser implementation of the pre-commit cli. I brought back the old hook and renamed the new one not to contain version in it. I also adapted the tests for original cli for the pre-commit cli. I added pycache files to .gitignore since I commited them accidentally (can revert this if you prefer it in a separate pr).

naglis commented on jsonschema/tests/test_pre_commit_cli.py:
```diff
@@ -0,0 +1,793 @@
+        error = str(e.exception)
+        self.assertIn(f"{os.sep}someNonexistentFile.json'", error)
+
+    def test_invalid_exlicit_base_uri(self):
```
```suggestion
    def test_invalid_explicit_base_uri(self):
```

AleksaC replied:
This typo is also present in the test for official cli. You can submit a PR with a fix if you want to.

janosh commented:
Feels like a lot of work went into this already. Anything I can do help get this in?

Julian commented:
So sorry, somehow I didn't realize this was waiting on me! Will have another look, though will likely take a few days to clear some other things off.

ssbarnea commented:
Any chance to revive this PR?

sirosen commented:
Per #910, we're going to start treating `check-jsonschema` as the official CLI for `jsonschema`. One of the responsibilities of that project is providing a set of pre-commit hooks. The generic `check-jsonschema` hook is comparable to the behavior of this PR.

I would recommend closing this. I know that people have worked hard on it -- and I don't feel great about recommending that the work be discarded -- but I believe that there's a better future for pre-commit hooks in a separate project from `jsonschema` itself. If there are valuable behaviors here which are missing from `check-jsonschema`, I'm happy to take PRs, feature requests, bug reports, or constructive criticism!

Julian commented:
> I would recommend closing this. I know that people have worked hard on it -- and I don't feel great about recommending that the work be discarded -- but I believe that there's a better future for pre-commit hooks in a separate project from jsonschema itself.

It's my apology to make, I'm sorry all for not getting to this :( some other (non-jsonschema things) have been sapping even more time away.

But yes agreed, we're going to move this functionality there, so please give feedback there.
